librz/type/parser/c_cpp_parser.c

```diff
@@ -0,0 +1,251 @@
+	//
+	// And only after that - run the normal C/C++ syntax parsing
+
+	// Filter types function prototypes and start parsing
+	int i = 0, result = 0;
+	for (i = 0; i < root_node_child_count; i++) {
+		parser_debug(state, "Processing %d child...\n", i);
+		TSNode child = ts_node_named_child(root_node, i);
+		result += parse_type_nodes_save(state, child, code);
+	}
+
+	// If there were errors during the parser then the result is different from 0
+	if (result) {
+		const char *error_msgs = rz_strbuf_drain_nofree(state->errors);
+		eprintf("Errors:\n");
+		eprintf(error_msgs);
+		const char *warning_msgs = rz_strbuf_drain_nofree(state->warnings);
+		eprintf("Warnings:\n");
+		eprintf(warning_msgs);
+		*error_msg = strdup(error_msgs);
+	}
+	if (state->verbose) {
+		const char *debug_msgs = rz_strbuf_drain_nofree(state->debug);
+		eprintf(debug_msgs);
+	}
+
+	// After everything parsed, we should preserve the base type database
+	c_parser_state_free_keep_ht(state);
+	ts_tree_delete(tree);
+	ts_parser_delete(parser);
+	return result;
+}
+
+RZ_API int rz_type_parse_string_stateless(RzTypeParser *parser, const char *code, char **error_msg) {
+	return type_parse_string(parser->state, code, error_msg);
+}
+
+RZ_API int rz_type_parse_file_stateless(RzTypeParser *parser, const char *path, const char *dir, char **error_msg) {
+	size_t read_bytes = 0;
+	const char *source_code = rz_file_slurp(path, &read_bytes);
+	if (!source_code || !read_bytes) {
+		return -1;
+	}
+	ut64 file_size = rz_file_size(path);
+	RZ_LOG_DEBUG("File size is %" PFMT64d " bytes, read %zu bytes\n", file_size, read_bytes);
+	return rz_type_parse_string_stateless(parser, source_code, error_msg);
+}
+
+RZ_API int rz_type_parse_file(RzTypeDB *typedb, const char *path, const char *dir, char **error_msg) {
+	size_t read_bytes = 0;
+	const char *source_code = rz_file_slurp(path, &read_bytes);
+	if (!source_code || !read_bytes) {
+		return -1;
+	}
+	ut64 file_size = rz_file_size(path);
+	RZ_LOG_DEBUG("File size is %" PFMT64d " bytes, read %zu bytes\n", file_size, read_bytes);
+	return rz_type_parse_string(typedb, source_code, error_msg);
+}
+
+RZ_API int rz_type_parse_string(RzTypeDB *typedb, const char *code, char **error_msg) {
+	bool verbose = true;
+	// Create new C parser state
+	CParserState *state = c_parser_state_new(typedb->types);
+	if (!state) {
+		eprintf("CParserState initialization error!\n");
+		return -1;
+	}
+	state->verbose = verbose;
+	return type_parse_string(state, code, error_msg);
+}
+
+RZ_API void rz_type_parse_reset(RzTypeDB *typedb) {
+	rz_type_parser_free(typedb->parser);
+	typedb->parser = rz_type_parser_new();
+}
+
+// Parses only single statement (the first one) and ignores everything else
+RZ_API RZ_OWN RzType *rz_type_parse_string_single(RzTypeParser *parser, const char *code, char **error_msg) {
+	// Create a parser.
+	TSParser *tsparser = ts_parser_new();
+	// Set the parser's language (C in this case)
+	ts_parser_set_language(tsparser, tree_sitter_c());
```

Comment:
Here we would use the second grammar for types only, right?

Reply:
Yes, this thing can be extended for more or slightly different grammars.

Reply:
Fixed by using slightly patched grammar.